### DISCUSSION TITLE
Cascading is done during Session.Save and Commit even though cascade attribute is not set.

### DISCUSSION THREAD

**Discussion by nishantgupta204** [nhibernate/nhibernate-core#2781] [Q&A]
Hi There,

In our application when I am trying to save the new created object, there is cascading happening before and after saving to the properties of the object which are of collection type and entity type. I have not specified cascade attribute in my HBM. When I look at the code for Session.Save, there is mandatory cascading happening. This is causing performance bottlenecks when I call Session.Save. Is there any way we can avoid this code block, as the relationship and order in which objects are added is handled by application itself.

Please reply ASAP as we have lot of performance bottlenecks coming due to this

**Comment by bahusoid** [comment]
> When I look at the code for Session.Save, there is mandatory cascading happening

Where exactly are you looking? Did you check the actual cascading code - all checks are there:

https://github.com/nhibernate/nhibernate-core/blob/6ab71d08b6c7f0a9c95895e94700b466b0743d6e/src/NHibernate/Engine/Cascade.cs#L107-L126

So if those `persister.HasCascades` and `style.DoCascade` checks don't work properly for you - please provide reproducible example.

**Reply by nishantgupta204** [reply]
Hi @bahusoid 

Below is the code
#https://github.com/nhibernate/nhibernate-core/blob/2bd3c1680e105dc496a4c97f6d8630ca362feee8/src/NHibernate/Event/Default/AbstractFlushingEventListener.cs#L187-L202




Here the iteration is going on for all the object in the PersistentContext( irrespective of whether the entities are modified or not) and checking for cascade.

**Reply by bahusoid** [reply]
>  irrespective of whether the entities are modified or not

You original concern was about cascading attributes: `Cascading is done even though cascade attribute is not set`

And now you are saying about modified entities.  So what exactly is your issue? Is it really related to cascading attributes? `CascadeOnFlush` call should be no op if no cascading attributes are defined for entity - so I don't see any issues so far.

**Reply by nishantgupta204** [reply]
Hi,

My concern was this whole loop should be avoided when there is no cascade. Cascade check is inside the block 'CascadeOnFlush'. It should have been on top for loop as a default implementation. Is there any specific reason to do this? 

Right now I have fixed it having custom implementation as suggested by now.

Thanks

**Reply by bahusoid** [reply]
I don't get how "no cascade" case can be determined outside of the loop. But feel free to provide PR with suggested fix.

**Comment by bahusoid** [comment]
> Is there any way we can avoid this code block

It looks like you need some custom flushing logic. You can implement it in your own custom class implementing `IFlushEventListener` (or by overriding `DefaultFlushEventListener` or `AbstractFlushingEventListener`). And then use it:

```C#
//Somewhere before session factory is created:
configuration.EventListeners.FlushEventListeners = new[]{ new  YourCustomFlushEventListener()};
```

Or in xml:
```xml
<hibernate-configuration>
    <session-factory>
        ...
        <event type="flush">
            <listener class="YourCustomFlushEventListener"/>
        </event>
    </session-factory>
</hibernate-configuration>
```
